src/xunit.v3.templates/templates/project/xunit3-extension/fs/SupportedOSAttribute.fs

```diff
@@ -37,16 +37,19 @@ type SupportedOSAttribute([<ParamArray>]supportedOSes: SupportedOS[]) =
 
         for supportedOS in supportedOSes do
             let osPlatform =
-                match osMappings.TryGetValue supportedOS with
-                | true, value -> value
-                | _ -> raise (new ArgumentException($"Supported OS value '{supportedOS}' is not a known OS", nameof(supportedOSes)))
+                match Map.tryFind supportedOS osMappings with
+                | Some value -> value
+                | None -> invalidArg (nameof supportedOSes) $"Supported OS value '{supportedOS}' is not a known OS"
 
-            if RuntimeInformation.IsOSPlatform osPlatform then matched = true |> ignore else () 
+            if RuntimeInformation.IsOSPlatform osPlatform then
+                matched <- true
+            else
+                () 
 
             if not matched then
```

Comment:
this check needs to be unindented to make any kind of sense - as it stands it will throw on the first iteration unless L45 runs. My guess it needs to become:
```
        let mutable matched = false

        for supportedOS in supportedOSes do
            match osMappings |> Map.tryFind supportedOS with
            | None -> invalidArg (nameof supportedOSes) $"Supported OS value '{supportedOS}' is not a known OS"
            | Some osPlatform ->
            
            if RuntimeInformation.IsOSPlatform osPlatform then
                matched <- true

          if not matched then
              // We use the dynamic skip exception message pattern to turn this into a skipped test
              // when it's not running on one of the targeted OSes
              failwith $"$XunitDynamicSkip$This test is not supported on {RuntimeInformation.OSDescription}"
```
NOTE I have unindented the `if not matched` line

If the goal is not intuitive understandability to someone coming from C#, I'd write:

```
   let targetPlatform (x: SupportedOS) =
        match osMappings |> Map.tryFind x with
        | None -> invalidArg (nameof supportedOSes) $"Supported OS value '{supportedOS}' is not a known OS"
        | Some target -> target
   let supportedHere = supportedOses |> Seq.map targetPlatform |> Seq.exists RuntimeInformation.IsOSPlatform 
   if not supportedHere then
       // We use the dynamic skip exception message pattern to turn this into a skipped test
       // when it's not running on one of the targeted OSes
       failwith $"$XunitDynamicSkip$This test is not supported on {RuntimeInformation.OSDescription}"
```

Reply:
I noticed the indentation error after merging this PR and have already fixed it.

Reply:
You can't move validation to the constructor of an attribute because of the way attributes are constructed. It's an oddity of the way attributes work.

Reply:
I like this iteration, and in my head I can see the C# LINQ version that this represents.

Reply:
Thanks, I've integrated your feedback: https://github.com/xunit/xunit/commit/69176a445eef78fc0098b8a16579aefb257cd651

Reply:
Thanks!

>You can't move validation to the constructor of an attribute because of the way attributes are constructed. It's an oddity of the way attributes work.

I kinda knew that; I was more trying to convey that I'd want to have the logic about the exception from parsing be separated out for clarity. Something like:

```
type SupportedOs with
    member x.ToOsPlatform() =
        match x with
        | SupportedOS.FreeBSD ->   OSPlatform.Create("FreeBSD")
        | SupportedOS.Linux ->     OSPlatform.Linux
        | SupportedOS.macOS ->     OSPlatform.OSX
        | SupportedOS.Windows ->   OSPlatform.Windows
        // TOCONSIDER isnt there another ctor that takes the value too, and then use a generic message?
        | x ->                     raise <| System.ArgumentOutOfRangeException(nameof x, $"Supported OS value '{x}' is not a known OS")

type SupportedOSAttribute([<ParamArray>]supportedOSes: SupportedOS[]) =
    inherit BeforeAfterTestAttribute()
    let supportedPlatforms = supportedOses |> Seq.map _.ToOsPlatform 
    override _.Before(_: MethodInfo, _: IXunitTest) =
        let canRunHere = supportedPlatforms |> Seq.exists RuntimeInformation.IsOSPlatform
        if not canRunHere then
           // We use the dynamic skip exception message pattern to turn this into a skipped test
           // when it's not running on one of the targeted OSes
           failwith $"$XunitDynamicSkip$This test is not supported on {RuntimeInformation.OSDescription}"

        Unchecked.defaultof<ValueTask>
```

Or can wrap it into the enum type:
```
type SupportedOS =
    | FreeBSD = 1
    | Linux = 2
    | macOS = 3
    | Windows = 4
    member x.ToOsPlatform() =
        match x with
        | SupportedOS.FreeBSD ->   OSPlatform.Create("FreeBSD")
        | SupportedOS.Linux ->     OSPlatform.Linux
        | SupportedOS.macOS ->     OSPlatform.OSX
        | SupportedOS.Windows ->   OSPlatform.Windows
        | x ->                     raise <| System.ArgumentOutOfRangeException(nameof x, $"Supported OS value '{x}' is not a known OS")
```
which would change 
> `let supportedPlatforms = supportedOses |> Seq.map SupportedOs.toOsPlatform`

to `let supportedPlatforms = supportedOses |> Seq.map _.ToOsPlatform`

Anyway, this is all subjective so will leave it up to you - there's definitely value in the three versions following the same outline and I'm already happy as it is ;)

Reply:
@bradwilson Thanks for the comments. I did unindent the change and failed to push the commit before running off to work!